tao_compiler/mlir/disc/transforms/disc_fuse_splat_const.cc

```diff
@@ -0,0 +1,118 @@
+/* Copyright 2021 The TensorFlow Authors. All Rights Reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+
+// This pass works after the DiscFusion pass, fuse the splat constant into its
+// consumers. The splat const will be duplicated if multiple consumers exist.
+// This pass can be regarded as a restricted version of the FusionMerger pass
+// of XLA, dealing with splat const as producer only.
+
+#include "mlir-hlo/Dialect/lhlo/IR/lhlo_ops.h"
+#include "mlir/Dialect/MemRef/IR/MemRef.h"
+#include "mlir/IR/Attributes.h"
+#include "mlir/IR/Builders.h"
+#include "mlir/IR/BuiltinTypes.h"
+#include "mlir/IR/Location.h"
+#include "mlir/IR/MLIRContext.h"
+#include "mlir/Pass/Pass.h"
+#include "tensorflow/compiler/mlir/disc/transforms/PassDetail.h"
+
+namespace mlir {
+namespace disc_ral {
+
+using namespace lmhlo;
+
+namespace {
+
+class DiscFuseSplatConstPass
+    : public DiscFuseSplatConstPassBase<DiscFuseSplatConstPass> {
+  void runOnOperation() override;
+
+ private:
+  void processSplatConst(
+      lmhlo::ConstOp constant,
+      DenseMap<lmhlo::FusionOp, SmallVector<Operation*, 4>> users);
+};
+
+// step 1, find all the unfused splat constants
+// step 2, fuse them in to the consumers
+// step 3, if not used anymore, erase the original
+void DiscFuseSplatConstPass::runOnOperation() {
+  FuncOp func = getOperation();
+  auto* context = &this->getContext();
+
+  SmallVector<lmhlo::ConstOp, 4> worklist;
+  func.walk([&](lmhlo::ConstOp constant) {
+    if (constant.value().isSplat() &&
+        (constant->getParentOfType<lmhlo::FusionOp>() == nullptr)) {
```

Comment:
disc cpu stitch fusion currently can not fuse const a.t.m. So we may need to skip stitch fusion op on cpu.

Reply:
why not for cpu stitch fusion?
existing tensorflow_ops already covers such cases (e.g. floor_mod.cc) on cpu backend, and the result is OK.
Any additional considerations on performance?